torchtune/generation/_generation.py

```diff
@@ -69,86 +100,213 @@ def update_stop_tokens_tracker(
         >>> tokenizer = torchtune.models.llama3.llama3_tokenizer()
-        >>> prompt = tokenizer("Hi my name is")
-        >>> output = generate(model, prompt, max_generated_tokens=100)
-        >>> print(tokenizer.decode(output[0]))
+        >>> prompt = tokenizer.encode("Hi my name is")
+        >>> rng.manual_seed(42)
+        >>> output, logits = generate(model, torch.tensor(prompt), max_generated_tokens=100, pad_id=0)
+        >>> print(tokenizer.decode(output[0].tolist()))
         Hi my name is Jeremy and I'm a friendly language model assistant!
 
     Returns:
-        List[List[int]]: collection of lists of generated tokens
+        torch.Tensor: Generated tokens.
     """
     prompt = prompt.view(1, -1) if prompt.ndim == 1 else prompt
-    # convert stop tokens to tensor for easy matching
+
     stop_tokens = (
         torch.tensor(stop_tokens, device=prompt.device) if stop_tokens else None
     )
+
+    if custom_generate_next_token is None:
+        custom_generate_next_token = generate_next_token
+
     bsz, prompt_length = prompt.size()
+    total_response_length = prompt_length + max_generated_tokens
+
     generated_tokens = prompt.clone()
-    # keeps track at a high level if we've already hit a stop token in a sequence so we can early stop
-    stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
-    # everything in stop_token_mask starts as 1s, and we'll set them to 0 for sequences
-    # that already hit a stop token
-    stop_token_mask = torch.ones(
-        (bsz, prompt_length + 1), dtype=torch.int32, device=prompt.device
+    incremental_decoding = model.caches_are_enabled()
+    max_seq_len = (
+        total_response_length if not incremental_decoding else model.max_seq_len
     )
+    padding_masks = generated_tokens != pad_id
+    cache_pos = None
 
-    if custom_generate_next_token is None:
-        custom_generate_next_token = generate_next_token
+    if not padding_masks.all():
+        padding_masks = torch.nn.functional.pad(
+            padding_masks, (0, max_generated_tokens), value=True
+        )
+        masks = get_causal_mask_from_padding_mask(
+            padding_masks, target_seq_len=max_seq_len
+        )
+        input_pos = get_position_ids_from_padding_mask(padding_masks)
+        cache_pos = torch.arange(
+            0, total_response_length, device=generated_tokens.device
+        )
+    else:
+        masks = torch.tril(
+            torch.ones(
+                total_response_length,
+                max_seq_len,
+                dtype=torch.bool,
+                device=prompt.device,
+            )
+        ).unsqueeze(0)
+        input_pos = torch.arange(
+            0, total_response_length, device=generated_tokens.device
+        ).unsqueeze(0)
+
+    del padding_masks
+
+    if incremental_decoding:
+        curr_masks = masks[:, :prompt_length]
+    else:
+        curr_masks = masks[:, :prompt_length, :prompt_length]
```

Comment:
Sorry I'm having a hard part following this. Can you add some comments here explaining what's happening?